### PR TITLE
rpc: fundrawtransaction and walletcreatefundedpsbt also lock manually selected coins

### PR DESCRIPTION
**Pull request by Sjors**
When using `fundrawtransaction` and `walletcreatefundedpsbt` with `lockUnspents`, it would only lock automatically selected coins, not manually selected coins. That doesn't make much sense to me if the goal is to prevent accidentally double-spending yourself before you broadcast a transaction.

Note that when  creating a transaction, manually selected coins are automatic "unlocked" (or more accurately: the lock is ignored). Earlier versions of this PR introduced an error when a locked coin is manually selected, but this idea was abandoned after some discussion. An application that uses this RPC should either rely on automatic coin selection (with `lockUnspents`) or handle lock concurrency itself with manual coin selection. In particular it needs to make sure to avoid/pause calls with automatic coin selection between calling `lockunspent` and the subsequent spending RPC.    

See #7518 for historical background.

### CODE REVIEW

**Comment by promag** [comment]
Concept ACK!

This made me think that we should fail funding if some coin is already locked?

File: test/functional/rpc_psbt.py [hunk 1]
```diff
@@ -89,6 +89,15 @@ def run_test(self):
         final_tx = self.nodes[0].finalizepsbt(signed_tx)['hex']
         self.nodes[0].sendrawtransaction(final_tx)
 
+        # Manually selected inputs can be locked:
+        utxo1 = self.nodes[0].listunspent()[0]
+        psbtx1 = self.nodes[0].walletcreatefundedpsbt([{"txid": utxo1['txid'], "vout": utxo1['vout']}], {self.nodes[2].getnewaddress():1}, 0,{"lockUnspents": True})["psbt"]
+        assert_equal(len(self.nodes[0].listlockunspent()), 1)
```
Comment (promag) [inline test/functional/rpc_psbt.py]: `assert_equal(len(self.nodes[0].listlockunspent()), 0)` before `walletcreatefundedpsbt`?

**Review by promag — COMMENTED** [review]
Add a release note maybe? Clients unaware of this change can have dangling utxos locked.

**Comment by Sjors** [comment]
Added a release note.

> This made me think that we should fail funding if some coin is already locked?

Maybe in a separate PR. I also think it's safer for users to explicitly unlock coins before using them.

**Comment by promag** [comment]
> I also think it's safer for users to explicitly unlock coins before using them.

The unspent locking was done to deal with concurrent clients and so a client shouldn't take the decision to unlock a coin so lightly - it might screw with a concurrent funding. This is a possible use case and for this reason the user should be able to lock all or nothing IMO.

**Comment by Sjors** [comment]
>  the lock-an-already-locked-coin-and-ignore behavior

Not sure what you mean. This PR can potentially lock a coin that's already locked, but why is that a problem? This PR does not ignore new things.

**Comment by promag** [comment]
The problem is that you might screw with a concurrent funding. When you lock a coin you should expect that it won't be used by others.

**Comment by Sjors** [comment]
If a concurrent wallet consumer uses manual coin selection, I don't think this PR makes their problem worse. If it uses automatic automatic coin selection then this PR has no impact.

**Comment by Sjors** [comment]
After [IRC discussion](http://www.erisian.com.au/bitcoin-core-dev/log-2020-03-06.html#l-288) I changed this PR to honour locks for manually selected coins.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -2150,6 +2150,7 @@ static UniValue lockunspent(const JSONRPCRequest& request)
                 "Temporarily lock (unlock=false) or unlock (unlock=true) specified transaction outputs.\n"
                 "If no transaction outputs are specified when unlocking then all current locked transaction outputs are unlocked.\n"
                 "A locked transaction output will not be chosen by automatic coin selection, when spending bitcoins.\n"
+                "With manual coin selection a locked transaction output must unlocked before use."
```
Comment (promag) [inline src/wallet/rpcwallet.cpp]: "... must be ..."

File: test/functional/wallet_basic.py [hunk 3]
```diff
@@ -136,11 +136,20 @@ def run_test(self):
                                 self.nodes[2].lockunspent, False,
                                 [{"txid": unspent_0["txid"], "vout": 999}])
 
-        # An output should be unlocked when spent
+        # The lock on a manually selected output is honoured
         unspent_0 = self.nodes[1].listunspent()[0]
         self.nodes[1].lockunspent(False, [unspent_0])
         tx = self.nodes[1].createrawtransaction([unspent_0], { self.nodes[1].getnewaddress() : 1 })
-        tx = self.nodes[1].fundrawtransaction(tx)['hex']
```
Comment (promag) [inline test/functional/wallet_basic.py]: Are you removing this case because it's used elsewhere?

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -2453,15 +2453,19 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nC
 
     coinControl.fAllowOtherInputs = true;
 
+    // Checking existing locks and acquiring new locks (when lockUnspents is true)
+    // is an atomic operation.
+    auto locked_chain = chain().lock();
+    LOCK(cs_wallet);
+
     for (const CTxIn& txin : tx.vin) {
+        if(IsLockedCoin(txin.prevout.hash, txin.prevout.n)) {
```
[the remark this answers is not in the document]
Comment (Sjors) [reply src/wallet/wallet.cpp]: The final frontier.

File: test/functional/wallet_basic.py (hunk 3 above)
Comment (Sjors) [reply test/functional/wallet_basic.py]: The original test was for auto-unlocking a coin, which now produces an error, so I check for that.

File: test/functional/wallet_basic.py [hunk 5]
```diff
@@ -136,11 +136,20 @@ def run_test(self):
...
+        assert_raises_rpc_error(-4, "Output " + unspent_0['txid'] + ":" + str(unspent_0['vout']) + " is locked",
+                                self.nodes[1].fundrawtransaction,tx,{"lockUnspents": True})
+
+        # fundrawtransaction can lock an input
+        self.nodes[1].lockunspent(True, [unspent_0])
+        assert_equal(len(self.nodes[1].listlockunspent()), 0)
+        tx = self.nodes[1].fundrawtransaction(tx,{"lockUnspents": True})['hex']
```
Comment (promag) [inline test/functional/wallet_basic.py]: nit, add space if you happen to push again.

**Review by promag — COMMENTED** [review]
Code review ACK 991a38ce9e9c19ec78139d2988b2abd7c9983519.

**Comment by instagibbs** [comment]
concept ACK

**Review by meshcollider — COMMENTED** [review]
utACK 991a38ce9e9c19ec78139d2988b2abd7c9983519

**Comment by Sjors** [comment]
Rebased

**Comment by achow101** [comment]
ACK 74992fd8cbe6f409bd1dad6c064deb5428509472

**Review by promag — COMMENTED** [review]
> This made me think that we should fail funding if some coin is already locked?

After reading past discussions and some more careful analysis I now think this is not correct.

With the current approach, a client C1 must unlock the unspents to manually use them. But right after `lockunspent true ...` some other client C2 can fund/send and and take C1 unspents, and C1 will error.

So I think the "is locked" check should be used only for automatic coin selection, not for those explicitly set.

This means that:
 a) the client handles concurrency elsewhere and doesn't use locked unspents
 b) the client locks unspents beforehand and then use them by funding transactions with those as inputs - and also sets `lockUnspents` to grab the new ones
 c) the user makes consecutive fundings with `lockUnspents` set.
 
In other words, we can assume that the client that funds a transaction with locked inputs is the one that locked them in the first place.

The c) case above is relevant - with the current approach it wouldn't be possible because between each funding iteration all inputs would have to be unlocked.

Lastly, the option `lockUnspents` when `true` should always lock all inputs - the caller will "own" all inputs (existing inputs whether locked or not and the new ones). This was the original change.

Sorry @Sjors for the misdirection 😞

**Review by jonatack — COMMENTED** [review]
Almost-ACK, review/tested, modulo I need to think about https://github.com/bitcoin/bitcoin/pull/18244#pullrequestreview-430047339.

**Comment by Sjors** [comment]
I'll await the above discussion. Will address nits if that leads to a change.

**Comment by Sjors** [comment]
Based on the above discussion I changed this PR again, it now does the following:
1. lock all inputs for a transaction, rather than just the automatically selected ones
2. unchanged behaviour:  ignore lock on manually coins (with some tests)

**Comment by promag** [comment]
Thanks @Sjors will review again.

**Review by meshcollider — COMMENTED** [review]
Code review ACK 6d1f51343cf11b07cd401fbd0c5bc3603e185a0e

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -2047,6 +2047,7 @@ static UniValue lockunspent(const JSONRPCRequest& request)
                 "Temporarily lock (unlock=false) or unlock (unlock=true) specified transaction outputs.\n"
                 "If no transaction outputs are specified when unlocking then all current locked transaction outputs are unlocked.\n"
                 "A locked transaction output will not be chosen by automatic coin selection, when spending bitcoins.\n"
+                "Manually selected coins are automatically unlocked.\n"
```
Comment (promag) [inline src/wallet/rpcwallet.cpp:2050]: This is incorrect, or am I missing something?

**Review by promag — COMMENTED** [review]
Just want to point out that it's impossible to prevent 2 clients spend the same manually specified inputs.

One way to fix this is to make each client explicitly lock those inputs, and only proceed if the lock was granted.

File: src/wallet/rpcwallet.cpp (hunk 6 above)
Comment (Sjors) [reply src/wallet/rpcwallet.cpp:2050]: Yes, but that's not the point. Automatically selection avoids locks, manually selection causes an unlock. This remark projects a user who assumes locked coins can't be accidentally spend by manually picking them.

**Comment by Sjors** [comment]
Client-specific locks sounds fun, but not in the PR :-)

File: src/wallet/rpcwallet.cpp (hunk 6 above)
Comment (fjahr) [reply src/wallet/rpcwallet.cpp:2050]: I think "Locks on manually selected coins are ignored." would be a more intuitive description of what happens but I understand this to mean the same thing.

**Review by fjahr — COMMENTED** [review]
Code review ACK 6d1f51343cf11b07cd401fbd0c5bc3603e185a0e
